### PR TITLE
Add PSNR API

### PR DESCRIPTION
**Pull request by IMvision12**
@fchollet

### CODE REVIEW

File: keras/src/backend/tensorflow/image.py [hunk 1]
```diff
@@ -15,6 +16,18 @@
...
 
+def psnr(image1, image2, max_val):
+    if image1.shape != image2.shape:
+        raise ValueError(
+            f"Image shapes {image1.shape} and {image2.shape} must match for PSNR calculation."
+        )
+
+    max_val = convert_to_tensor(max_val, dtype=image1.dtype)
+    mse = tf.reduce_mean(tf.square(image1 - image2), axis=[-3, -2, -1])
+    psnr_val = 20 * tfnp.log10(max_val) - 10 * tfnp.log10(mse)
```
Comment (james77777778) [inline keras/src/backend/tensorflow/image.py:27]: Hi @IMvision12 
I'm trying to remove the dependency on `tfnp` in #19556 

I think you could use something like this in the `psnr` function to prevent future breakage, and it should be more consistent regarding dtype promotion.
```python
from keras.src.backend.tensorflow.numpy import log10
```

Comment (IMvision12) [reply keras/src/backend/tensorflow/image.py:27]: @james77777778 Ok, That's cool!

Comment (IMvision12) [reply keras/src/backend/tensorflow/image.py]: @james77777778 any idea on failing tests?

File: keras/src/backend/tensorflow/image.py [hunk 2]
```diff
@@ -15,6 +15,19 @@
 )
 
 
+def psnr(image1, image2, max_val):
+    from keras.src.backend.tensorflow.numpy import log10
+    if image1.shape != image2.shape:
+        raise ValueError(
+            f"Image shapes {image1.shape} and {image2.shape} must match for PSNR calculation."
```
Comment (fchollet) [inline keras/src/backend/tensorflow/image.py:22]: Reproduce the same error message as in the other functions.

File: keras/src/ops/image.py [hunk 3]
```diff
@@ -7,6 +7,78 @@
...
+
+
+@keras_export("keras.ops.image.psnr")
+def psnr(
+    image1,
+    image2,
+    max_val,
+):
+    """Peak Signal-to-Noise Ratio (PSNR) calculation.
+    This function calculates the Peak Signal-to-Noise Ratio between two images
```
Comment (fchollet) [inline keras/src/ops/image.py:46]: Add line breaks between sections

File: keras/src/backend/jax/image.py [hunk 4]
```diff
@@ -14,6 +14,20 @@
 )
 
 
+def psnr(image1, image2, max_val):
+    if image1.shape != image2.shape:
+        raise ValueError(
+            "Image shapes must match for PSNR calculation. "
```
Comment (fchollet) [inline keras/src/backend/jax/image.py:20]: The error message should not reference images -- any signal should be valid.

File: keras/src/backend/jax/image.py [hunk 5]
```diff
@@ -14,6 +14,20 @@
...
+            f"Received shapes: image1.shape={image1.shape}, "
+            "image2.shape={image2.shape}. "
+        )
+
+    max_val = convert_to_tensor(max_val, image1.dtype)
+    mse = jnp.mean(jnp.square(image1 - image2), axis=(-3, -2, -1))
```
Comment (fchollet) [inline keras/src/backend/jax/image.py:26]: The intent of `axis=(-3, -2, -1)` is apparently to exclude the batch axis. But is that necessary? The TF API has no such behavior and returns a scalar.

File: keras/src/ops/image.py [hunk 6]
```diff
@@ -7,6 +7,78 @@
...
+            max_val=self.max_val,
+        )
+
+    def compute_output_spec(self, image1, image2):
+        if len(image1.shape) != len(image2.shape):
+            raise ValueError("Input images must have the same rank")
+
+        if len(image1.shape) not in (3, 4):
+            raise ValueError(
+                "Invalid image rank: expected rank 3 (single image) "
```
Comment (fchollet) [inline keras/src/ops/image.py:31]: There is nothing in the computation that relies on shapes. PSNR is something that is valid for any signal, not just images. So we should:

- Move this API to `nn` instead of `image`
- Remove all shape checks (except checking that the two input have the same shape)

**Review by fchollet — COMMENTED** [review]
Thanks for the update!

File: keras/src/backend/jax/image.py (hunk 5 above)
Comment (IMvision12) [reply keras/src/backend/jax/image.py:26]: Yup, will remove it!

File: keras/src/backend/tensorflow/image.py (hunk 2 above)
Comment (IMvision12) [reply keras/src/backend/tensorflow/image.py]: Ok

File: keras/src/backend/jax/image.py (hunk 5 above)
Comment (IMvision12) [reply keras/src/backend/jax/image.py]: When using axis=(1, 2, 3), the mean squared error (MSE) is computed across all dimensions except the batch dimension. If the user provides batched data, this ensures that PSNR will return a single value instead of one for each batch, which might not be desirable. Instead, it should return a PSNR value for each item in the batch, maintaining consistency with the batch size.

Comment (IMvision12) [reply keras/src/backend/jax/image.py]: TF API has this behavior

[PSNR](https://github.com/tensorflow/tensorflow/blob/5bc9d26649cca274750ad3625bd93422617eed4b/tensorflow/python/ops/image_ops_impl.py#L4211)

File: keras/src/backend/jax/nn.py [hunk 7]
```diff
@@ -659,3 +659,15 @@ def _iterate(prev, x):
...
+
+
+def psnr(x, y, max_val):
+    if x.shape != y.shape:
+        raise ValueError(
+            f"Expected x and y to have the same shape. Got x: {x.shape} and y: {y.shape}."
+        )
+
+    max_val = convert_to_tensor(max_val, x.dtype)
+    mse = jnp.mean(jnp.square(x - y), axis=(-3, -2, -1))
```
Comment (fchollet) [inline keras/src/backend/jax/nn.py:671]: We should just remove the axis argument -- right now it presupposes that the input has at least 3 axes, but in practice the input could have any rank >= 2.
